examples/lox/src/language/type-system/lox-type-checking.ts

```diff
@@ -0,0 +1,287 @@
+    // binary operators: numbers => number
+    for (const operator of ['-', '*', '/']) {
+        operators.createBinaryOperator({ name: operator, signature: { left: typeNumber, right: typeNumber, return: typeNumber }, inferenceRule: binaryInferenceRule });
+    }
+    operators.createBinaryOperator({ name: '+', signature: [
+        { left: typeNumber, right: typeNumber, return: typeNumber },
+        { left: typeString, right: typeString, return: typeString },
+        { left: typeNumber, right: typeString, return: typeString },
+        { left: typeString, right: typeNumber, return: typeString },
+    ], inferenceRule: binaryInferenceRule });
+
+    // TODO design decision: overload with the lowest number of conversions wins!
+    // TODO remove this later, it is not required for LOX!
+    // typir.conversion.markAsConvertible(typeNumber, typeString, 'IMPLICIT'); // var my1: string = 42;
+
+    // binary operators: numbers => boolean
+    for (const operator of ['<', '<=', '>', '>=']) {
+        operators.createBinaryOperator({ name: operator, signature: { left: typeNumber, right: typeNumber, return: typeBool }, inferenceRule: binaryInferenceRule });
+    }
+
+    // binary operators: booleans => boolean
+    for (const operator of ['and', 'or']) {
+        operators.createBinaryOperator({ name: operator, signature: { left: typeBool, right: typeBool, return: typeBool }, inferenceRule: binaryInferenceRule });
+    }
+
+    // ==, != for all data types (the warning for different types is realized below)
+    for (const operator of ['==', '!=']) {
+        operators.createBinaryOperator({ name: operator, signature: { left: typeAny, right: typeAny, return: typeBool }, inferenceRule: binaryInferenceRule });
+    }
+    // = for SuperType = SubType (TODO integrate the validation here? should be replaced!)
+    operators.createBinaryOperator({ name: '=', signature: { left: typeAny, right: typeAny, return: typeAny }, inferenceRule: binaryInferenceRule });
+
+    // unary operators
+    operators.createUnaryOperator({ name: '!', signature: { operand: typeBool, return: typeBool }, inferenceRule: unaryInferenceRule });
+    operators.createUnaryOperator({ name: '-', signature: { operand: typeNumber, return: typeNumber }, inferenceRule: unaryInferenceRule });
+
+    // define types which are declared by the users of LOX => investigate the current AST
+    const domainNodeRoot = AstUtils.getContainerOfType(domainNodeEntry, isLoxProgram)!;
+    AstUtils.streamAllContents(domainNodeRoot).forEach((node: AstNode) => {
+        // function types: they have to be updated after each change of the Langium document, since they are derived from FunctionDeclarations!
+        if (isFunctionDeclaration(node)) {
+            const functionName = node.name;
+            // define function type
+            functionKind.createFunctionType({
+                functionName,
+                outputParameter: { name: FUNCTION_MISSING_NAME, type: mapType(node.returnType) },
```

Comment:
FUNCTION_MISSING_NAME => RETURN_PARAMETER_NAME?

Reply:
Would you suggest to introduce `RETURN_PARAMETER_NAME` as another constant for this purpose?